File: src/main/java/my/Collector.java
package my;

import com.carrotsearch.hppc.LongObjectHashMap;
import com.carrotsearch.hppc.LongObjectMap;
import com.carrotsearch.hppc.LongScatterSet;
import com.carrotsearch.hppc.procedures.LongObjectProcedure;
import com.firenio.buffer.ByteBuf;
import com.firenio.codec.http11.HttpCodec;
import com.firenio.codec.http11.HttpFrame;
import com.firenio.codec.http11.HttpStatus;
import com.firenio.common.ByteUtil;
import com.firenio.common.Util;
import com.firenio.component.Channel;
import com.firenio.component.ChannelAcceptor;
import com.firenio.component.ChannelConnector;
import com.firenio.component.Frame;
import com.firenio.component.IoEventHandle;
import com.firenio.component.NioEventLoop;
import com.firenio.component.NioEventLoopGroup;
import com.firenio.component.ProtocolCodec;
import com.firenio.concurrent.Callback;
import com.firenio.log.DebugUtil;

import java.net.URLDecoder;
import java.security.MessageDigest;
import java.security.NoSuchAlgorithmException;
import java.util.Arrays;

import static my.Main.*;

/**
 * @author: wangkai
 **/
public class Collector {

    static final int                  TRACE_COUNT      = 5000;
    static final int                  MAX_ERR_TRACE    = 256;
    static final int                  CORE_SIZE        = 1;
    static final MessageDigest        MD5              = initMD5();
    static final NioEventLoopGroup    EL_GROUP         = new NioEventLoopGroup(true, CORE_SIZE);
    static final LongObjectMap<Trace> TRACE_MAP        = new LongObjectHashMap<>(1024 * 16);
    static final ByteBuf              RESULT           = ByteBuf.buffer(1024 * 1024);
    static final int[]                TRACE_POS_ARRAY  = new int[MAX_ERR_TRACE * TRACE_COUNT];
    static final long[]               TRACE_TIME_ARRAY = new long[MAX_ERR_TRACE * TRACE_COUNT];
    static final byte[]               TRACE_DATA       = new byte[MAX_TRACE_LEN * MAX_ERR_TRACE * TRACE_COUNT];
    static final LongScatterSet       TIME_SET         = new LongScatterSet(MAX_ERR_TRACE * 2 * TRACE_COUNT);

    static int          client_size;
    static ByteBuf      score_header;
    static boolean      ready;
    static int          r_port;
    static int          finish_1;
    static int          finish_2;
    static Channel      ch_score;
    static Channel      ch_8000_0;
    static Channel      ch_8000_1;
    static Channel      ch_8001_0;
    static Channel      ch_8001_1;
    static NioEventLoop event_loop;

    static void init(int port) throws Exception {
        RESULT.writeBytes("result=%7B".getBytes());
        EL_GROUP.setMemoryUnit(1024 * 64);
        EL_GROUP.setMemoryCapacity(1024 * 1024 * 256);
        EL_GROUP.setEventLoopName("nio-collector");
        EL_GROUP.setIdleTime(Long.MAX_VALUE);
        EL_GROUP.start();
        event_loop = EL_GROUP.getEventLoop(0);

        String req = "POST /api/finished HTTP/1.1\r\n" +
                "Host: localhost\r\n" +
                "Content-Type: application/x-www-form-urlencoded\r\n" +
                "Content-Length: ";
        score_header = ByteBuf.buffer(req.length() + 32);
        score_header.writeBytes(req.getBytes());
    }

    static void runCollector(int port) throws Exception {
        init(port);
        ChannelAcceptor http_context = new ChannelAcceptor(EL_GROUP, HOST, port);
        http_context.addChannelEventListener(NO_DELAY_LISTENER);
        http_context.addProtocolCodec(new HttpCodec("s", 0, true));
        http_context.setPrintConfig(false);
        http_context.setIoEventHandle(new IoEventHandle() {
            @Override
            public void accept(Channel ch, Frame frame) throws Exception {
                HttpFrame f   = (HttpFrame) frame;
                String    url = f.getRequestURL();
                if ("/setParameter".equals(url)) {
                    r_port = Integer.parseInt(f.getRequestParam("port"));
                    if (DEBUG) {
                        DebugUtil.info("set port: " + r_port);
                    }
                    connect_to_score(r_port);
                }
                DebugUtil.info("is_ready: " + ready);
                if (!ready) {
                    f.setStatus(HttpStatus.C503);
                }
                ch.writeAndFlush(frame);
            }
        });
        http_context.bind();
        ChannelAcceptor context = new ChannelAcceptor(EL_GROUP, HOST, C_PORT);
        context.addChannelEventListener(NO_DELAY_LISTENER);
        context.addProtocolCodec(new ProtocolCodec() {

            @Override
            public void read_buf(Channel ch, ByteBuf src) throws Exception {
                for (; ; ) {
                    int raw_len = src.readIntLE();
                    int len     = get_len_from_type_len(raw_len);
                    if (src.readableBytes() < len) {
                        src.skipRead(-4);
                        store_plain_remain(ch, src);
                        return;
                    }
                    int type = get_type(raw_len);
                    if (type == TYPE_ERR_ID) {
                        ByteBuf buf = ch.allocate(len + 4);
                        buf.writeIntLE(len);
                        buf.writeBytes(src, len);
                        handle_write_pair(ch, buf);
                    } else if (type == TYPE_ADD_WAIT) {
                        ByteBuf buf = ch.allocate(4);
                        buf.writeIntLE(raw_len);
                        handle_write_pair(ch, buf);
                    } else if (type == TYPE_ERR_DATA) {
                        src.markWriteIndex();
                        src.writeIndex(src.readIndex() + len);
                        long  trace_id = src.readLongLE();
                        Trace trace    = getTrace(trace_id);
                        trace.write(src);
                        src.resetWriteIndex();
                    } else if (type == TYPE_FINISH) {
                        int s_type = src.readIntLE();
                        if (s_type == 1) {
                            if (++finish_1 == 4) {
                                call_processor_finish();
                            }
                        } else {
                            if (++finish_2 == 4) {
                                handle_score_finish();
                            }
                        }
                    } else if (type == TYPE_SET_PORT) {
                        int port = src.readIntLE();
                        if (port == 8000_0) {
                            ch_8000_0 = ch;
                        } else if (port == 8000_1) {
                            ch_8000_1 = ch;
                        } else if (port == 8001_0) {
                            ch_8001_0 = ch;
                        } else if (port == 8001_1) {
                            ch_8001_1 = ch;
                        }
                        if (++client_size == 4) {
                            call_processor_warm_up();
                        }
                    } else {
                        log_and_exit("error read type: " + raw_len);
                    }
                    if (!src.hasReadableBytes()) {
                        return;
                    }
                }
            }
        });
        context.setPrintConfig(false);
        context.bind();
    }

    static void connect_to_score(int port) throws Exception {
        ChannelConnector conn = new ChannelConnector(event_loop, HOST, port);
        conn.addChannelEventListener(NO_DELAY_LISTENER);
        conn.addProtocolCodec(new ProtocolCodec() {
            @Override
            public void read_buf(Channel ch, ByteBuf src) throws Exception {
                if (OFFLINE) {
                    System.out.println(new String(src.readBytes(src.writeIndex())));
                }
            }
        });
        conn.setPrintConfig(false);
        conn.connect(new Callback<Channel>() {
            @Override
            public void call(Channel res, Throwable ex) {
                if (ex != null) {
                    Util.sleep(100);
                    try {
                        connect_to_score(port);
                    } catch (Exception exception) {
                        DebugUtil.error(exception.getMessage(), exception);
                    }
                } else {
                    ch_score = res;
                }
            }
        });

    }

    static void handle_write_pair(Channel ch, ByteBuf buf) {
        if (ch == ch_8000_0) {
            ch_8001_0.writeAndFlush(buf);
        } else if (ch == ch_8000_1) {
            ch_8001_1.writeAndFlush(buf);
        } else if (ch == ch_8001_0) {
            ch_8000_0.writeAndFlush(buf);
        } else if (ch == ch_8001_1) {
            ch_8000_1.writeAndFlush(buf);
        }
    }

    static void call_processor_warm_up() {
        if (OFFLINE) {
            DebugUtil.info("call_processor_warm_up");
        }
        ByteBuf buf = event_loop.alloc().allocate(4);
        buf.writeIntLE(update_type(0, TYPE_WARM_UP));
        ch_8000_0.writeAndFlush(buf.duplicate());
        ch_8001_0.writeAndFlush(buf);
    }

    static void call_processor_finish() {
        ByteBuf buf = event_loop.alloc().allocate(4);
        buf.writeIntLE(update_type(0, TYPE_FINISH));
        ch_8000_0.writeAndFlush(buf.duplicate());
        ch_8000_1.writeAndFlush(buf.duplicate());
        ch_8001_0.writeAndFlush(buf.duplicate());
        ch_8001_1.writeAndFlush(buf);
    }

    static void handle_score_finish() {
        if (ready) {
            TRACE_MAP.forEach((LongObjectProcedure<? super Trace>) (key, value) -> {
                append_result(value);
            });
            RESULT.skipWrite(-3);
            append_right_brace();
            if (OFFLINE) {
                String result  = new String(RESULT.getBytes(0, RESULT.writeIndex()));
                String decoded = URLDecoder.decode(result);
                System.out.println(decoded);
                System.out.println();
            }
            score_header.writeBytes(String.valueOf(RESULT.writeIndex()).getBytes());
            score_header.writeByte((byte) '\r');
            score_header.writeByte((byte) '\n');
            score_header.writeByte((byte) '\r');
            score_header.writeByte((byte) '\n');
            ch_score.writeAndFlush(score_header);
            ch_score.writeAndFlush(RESULT);
        } else {
            TRACE_MAP.clear();
            TIME_SET.clear();
            RESULT.writeIndex(10);
            Arrays.fill(TRACE_POS_ARRAY, 0);
            Arrays.fill(TRACE_DATA, (byte) 0);
            Arrays.fill(TRACE_TIME_ARRAY, 0);
            finish_1 = 0;
            finish_2 = 0;
            ready = true;
            System.gc();
            if (OFFLINE) {
                DebugUtil.info("warm up ready");
            }
        }
    }

    static Trace getTrace(long trace_id) {
        Trace trace = TRACE_MAP.get(trace_id);
        if (trace == null) {
            trace = new Trace();
            TRACE_MAP.put(trace_id, trace);
        }
        return trace;
    }

    static class Trace {

        static int index = 0;

        Trace() {
            this.p = index++;
            this.pt_off = p * MAX_ERR_TRACE;
            this.data_off = p * MAX_TRACE_LEN * MAX_ERR_TRACE;
        }

        final int p;
        final int data_off;
        final int pt_off;
        int    size;
        int    data_size;
        byte[] md5;
        int    id_len = 0;

        void write(ByteBuf buf) {
            if (id_len == 0) {
                id_len = buf.indexOf(SP, buf.readIndex() + 2, buf.writeIndex()) - buf.readIndex() - 2;
            }
            int pt_off    = this.pt_off;
            int data_off  = this.data_off;
            int data_size = this.data_size;
            int i         = size;
            for (; buf.hasReadableBytes(); ) {
                int len = buf.readShortLE();
                buf.readBytes(TRACE_DATA, data_off + data_size, len);
                long time = get_trace_time(TRACE_DATA, data_off + data_size + id_len + 1, 16);
                if (TIME_SET.contains(time)) {
                    continue;
                } else {
                    TIME_SET.add(time);
                }
                TRACE_TIME_ARRAY[pt_off + i] = to_trace_time(time, i);
                TRACE_POS_ARRAY[pt_off + i] = to_trace_pos(data_size, len);
                data_size += len;
                i++;
            }
            this.size = i;
            this.data_size = data_size;
            this.md5();
            if (DEBUG) {
                if (size > MAX_ERR_TRACE) {
                    log_and_exit("err_trace_overflow");
                }
            }
        }

        void md5() {
            MD5.reset();
            int pt_off = this.pt_off;
            int pt_end = pt_off + size;
            Arrays.sort(TRACE_TIME_ARRAY, pt_off, pt_end);
            for (int i = pt_off; i < pt_end; i++) {
                long raw_time = TRACE_TIME_ARRAY[i];
                int  index    = get_trace_index(raw_time);
                int  raw_pos  = TRACE_POS_ARRAY[pt_off + index];
                int  pos      = get_trace_pos(raw_pos);
                int  len      = get_trace_len(raw_pos);
                MD5.update(TRACE_DATA, data_off + pos, len);
            }
            md5 = MD5.digest();
        }

        String getText() {
            return new String(TRACE_DATA, data_off, data_size);
        }

    }

    static MessageDigest initMD5() {
        try {
            return MessageDigest.getInstance("MD5");
        } catch (NoSuchAlgorithmException e) {
            throw new RuntimeException(e);
        }
    }

    static long get_trace_time(byte[] data, int off, int len) {
        int  end  = off + len;
        long time = 0;
        for (int i = off; i < end; i++) {
            time = time * 10 + data[i] - '0';
        }
        return time;
    }

    static int to_trace_pos(int pos, int len) {
        return (pos << 16) | len;
    }

    static int get_trace_pos(int raw_pos) {
        return raw_pos >>> 16;
    }

    static int get_trace_len(int raw_pos) {
        return raw_pos & 0xffff;
    }

    static long to_trace_time(long time, long index) {
        return (time << 7) | index;
    }

    static long get_trace_time(long raw_time) {
        return raw_time >>> 7;
    }

    static int get_trace_index(long raw_time) {
        return (int) (raw_time & (0xff >>> 1));
    }

    static void append_result(Trace trace) {
//        =  %3D
//        {  %7B
//        "  %22
//        :  %3A
//        }  %7D
//        ,  %2C

        int    data_off = trace.data_off;
        int    end      = data_off + trace.id_len;
        byte[] md5      = trace.md5;
        append_double_q();
        for (int i = data_off; i < end; i++) {
            RESULT.writeByte(TRACE_DATA[i]);
        }
        append_double_q();
        append_colon();
        append_double_q();
        for (int i = 0; i < md5.length; i++) {
            String hex = ByteUtil.getHexString(md5[i]);
            RESULT.writeByte((byte) hex.charAt(0));
            RESULT.writeByte((byte) hex.charAt(1));
        }
        append_double_q();
        append_comma();
    }

    static void append_eq() {
        RESULT.writeByte((byte) '%');
        RESULT.writeByte((byte) '3');
        RESULT.writeByte((byte) 'D');
    }

    static void append_left_brace() {
        RESULT.writeByte((byte) '%');
        RESULT.writeByte((byte) '7');
        RESULT.writeByte((byte) 'B');
    }

    static void append_double_q() {
        RESULT.writeByte((byte) '%');
        RESULT.writeByte((byte) '2');
        RESULT.writeByte((byte) '2');
    }

    static void append_colon() {
        RESULT.writeByte((byte) '%');
        RESULT.writeByte((byte) '3');
        RESULT.writeByte((byte) 'A');
    }

    static void append_right_brace() {
        RESULT.writeByte((byte) '%');
        RESULT.writeByte((byte) '7');
        RESULT.writeByte((byte) 'D');
    }

    static void append_comma() {
        RESULT.writeByte((byte) '%');
        RESULT.writeByte((byte) '2');
        RESULT.writeByte((byte) 'C');
    }

}

File: go/tail-base/src/wk/Collector.go
package wk

import (
	"../intintmap"
	"crypto/md5"
	"encoding/binary"
	"fmt"
	"hash"
	"net"
	"net/http"
	"net/url"
	"sort"
	"strconv"
	"unsafe"
)

const TRACE_COUNT = 5000
const MAX_ERR_TRACE = 256

var TRACE_POS_ARRAY [MAX_ERR_TRACE * TRACE_COUNT]uint32
var TRACE_TIME_ARRAY [MAX_ERR_TRACE * TRACE_COUNT]int64
var TRACE_DATA [MAX_TRACE_LEN * MAX_ERR_TRACE * TRACE_COUNT]byte
var TRACE_MAP = make(map[int64]*Trace, 1024*16)
var RESULT [1024 * 1024]byte
var result_write_index = 0
var TIME_SET = intintmap.New(MAX_ERR_TRACE*2*TRACE_COUNT, 0.5)
var score_header [1024]byte
var score_header_write_index = 0
var processor_connected = 0
var finish_1 = 0
var finish_2 = 0
var trace_index = 0
var ch_score net.Conn
var ch_8000_0 net.Conn
var ch_8000_1 net.Conn
var ch_8001_0 net.Conn
var ch_8001_1 net.Conn

var hexes = []byte{'0', '1', '2', '3', '4', '5', '6', '7', '8', '9', 'A', 'B', 'C', 'D', 'E', 'F'}

func Run_collector(port int) {
	collector_init(port)
	go listener_collector_http(port)
	go listener_8003()
}

func collector_init(port int) {
	append_result('r')
	append_result('e')
	append_result('s')
	append_result('u')
	append_result('l')
	append_result('t')
	append_result('=')
	append_left_brace()

	var req = "POST /api/finished HTTP/1.1\r\n" +
		"Host: localhost\r\n" +
		"Content-Type: application/x-www-form-urlencoded\r\n" +
		"Content-Length: "
	copy(score_header[0:], req)
	score_header_write_index += len(req)
}

func listener_collector_http(port int) {
	http.HandleFunc("/ready", Handle_collector_http_ready)
	http.HandleFunc("/setParameter", Handle_collector_http_set_parameter)
	var addr = HOST + ":" + strconv.FormatInt(int64(port), 10)
	err := http.ListenAndServe(addr, nil)
	if err != nil {
		fmt.Println("服务器错误")
	}
}

func Handle_collector_http_ready(res http.ResponseWriter, req *http.Request) {
	if processor_connected == 5 {
		res.WriteHeader(200)
	} else {
		res.WriteHeader(503)
	}
}

func Handle_collector_http_set_parameter(res http.ResponseWriter, req *http.Request) {
	if ONLINE {
		port, err := strconv.ParseInt(req.URL.Query().Get("port"), 10, 32)
		if err != nil {
			log_and_exit(err)
		}
		res.WriteHeader(200)
		ch_score = *connect(port)
	}
}

func listener_8003() {
	var addr = HOST + ":" + strconv.FormatInt(int64(C_PORT), 10)
	listener, err := net.Listen("tcp", addr)
	if err != nil {
		log_and_exit(err)
		return
	}
	if OFFLINE {
		fmt.Println("服务器等待客户端建立连接...")
	}
	for true {
		conn, err := listener.Accept()
		if err != nil {
			fmt.Println("accept err:", err)
			return
		}
		go handle_8003(&conn)
	}
}

func handle_8003(_conn *net.Conn) {
	var _MD5 = md5.New()
	var MD5 = &(_MD5)
	var conn = *_conn
	if OFFLINE {
		fmt.Printf("客户端与服务器连接建立成功...r: %s, l: %s \n", conn.RemoteAddr(), conn.LocalAddr())
	}
	var buf_len = 1024 * 512
	var read_buf = make([]byte, buf_len)
	var write_buf = make([]byte, buf_len)
	var write_index = 0
	var read_index = 0
	for true {
		n, err := conn.Read(read_buf[write_index:])
		if err != nil {
			fmt.Println("read err:", err)
			return
		}
		write_index += n
		for true {
			var raw_len = binary.LittleEndian.Uint32(read_buf[read_index:])
			var data_len = get_len_from_type_len(raw_len)
			if write_index-read_index < data_len {
				if read_index > 0 {
					var remain = write_index - read_index
					copy(read_buf[0:], read_buf[read_index:write_index])
					read_index = 0
					write_index = remain
				}
				break
			}
			read_index += 4
			var _type = get_type_from_type_len(raw_len)
			if _type == TYPE_ERR_ID {
				var data = read_buf[read_index-4 : read_index+data_len]
				read_index += data_len
				if conn == ch_8000_0 {
					ch_8001_0.Write(data)
				} else if conn == ch_8000_1 {
					ch_8001_1.Write(data)
				} else if conn == ch_8001_0 {
					ch_8000_0.Write(data)
				} else if conn == ch_8001_1 {
					ch_8000_1.Write(data)
				}
			} else if _type == TYPE_ADD_WAIT {
				var data = read_buf[read_index-4 : read_index]
				if conn == ch_8000_0 {
					ch_8001_0.Write(data)
				} else if conn == ch_8000_1 {
					ch_8001_1.Write(data)
				} else if conn == ch_8001_0 {
					ch_8000_0.Write(data)
				} else if conn == ch_8001_1 {
					ch_8000_1.Write(data)
				}
			} else if _type == TYPE_ERR_DATA {
				var trace_id = int64(binary.LittleEndian.Uint64(read_buf[read_index:]))
				var trace = getTrace(trace_id)
				var fromIndex = read_index + 8
				var toIndex = read_index + data_len
				read_index = toIndex
				trace.write(read_buf[fromIndex:toIndex], MD5)
			} else if _type == TYPE_FINISH {
				var s_type = binary.LittleEndian.Uint32(read_buf[read_index:])
				read_index += 4
				if s_type == 1 {
					finish_1++
					if finish_1 == 4 {
						var data = write_buf[0:4]
						binary.LittleEndian.PutUint32(data, update_type(0, TYPE_FINISH))
						ch_8000_0.Write(data)
						ch_8000_1.Write(data)
						ch_8001_0.Write(data)
						ch_8001_1.Write(data)
					}
				} else {
					finish_2++
					if finish_2 == 4 {
						for _, trace := range TRACE_MAP {
							append_trace(trace)
						}
						result_write_index -= 3
						append_right_brace()
						var len_str = strconv.FormatInt(int64(result_write_index), 10)
						var len_bytes = (*[]byte)(unsafe.Pointer(&len_str))
						copy(score_header[score_header_write_index:], *len_bytes)
						score_header_write_index += len(*len_bytes)
						score_header[score_header_write_index] = '\r'
						score_header_write_index++
						score_header[score_header_write_index] = '\n'
						score_header_write_index++
						score_header[score_header_write_index] = '\r'
						score_header_write_index++
						score_header[score_header_write_index] = '\n'
						score_header_write_index++
						if OFFLINE {
							var str = bytes_to_string(RESULT[0:result_write_index])
							res, err := url.PathUnescape(*str)
							if err != nil {
								log_and_exit(err)
							}
							fmt.Println(res)
							str = bytes_to_string(score_header[0:score_header_write_index])
							fmt.Println(*str)
						}
						if ONLINE {
							ch_score.Write(score_header[0:score_header_write_index])
							ch_score.Write(RESULT[0:result_write_index])
						}
					}
				}
			} else if _type == TYPE_SET_PORT {
				var port = binary.LittleEndian.Uint32(read_buf[read_index:])
				read_index += 4
				if port == 80000 {
					ch_8000_0 = conn
					if OFFLINE {
						fmt.Println("set_port_8000_0")
					}
				} else if port == 80001 {
					ch_8000_1 = conn
					if OFFLINE {
						fmt.Println("set_port_8000_1")
					}
				} else if port == 80010 {
					ch_8001_0 = conn
					if OFFLINE {
						fmt.Println("set_port_8001_0")
					}
				} else if port == 80011 {
					ch_8001_1 = conn
					if OFFLINE {
						fmt.Println("set_port_8001_1")
					}
				}
				processor_connected++
				if processor_connected == 4 {
					if OFFLINE {
						fmt.Println("ready...")
					}
					var _temp = write_buf[0:4]
					binary.LittleEndian.PutUint32(_temp, update_type(0, TYPE_PROCESSOR_READY))
					ch_8000_0.Write(_temp)
					ch_8001_0.Write(_temp)
					processor_connected++
				}
			} else {
				log_and_exit("error read type: " + strconv.FormatInt(int64(raw_len), 10))
			}
			if read_index == write_index {
				read_index = 0
				write_index = 0
				break
			}
		}
	}
}

func append_trace(_trace *Trace) {
	//        =  %3D
	//        {  %7B
	//        "  %22
	//        :  %3A
	//        }  %7D
	//        ,  %2C
	var trace = *_trace
	var data_off = trace.data_off
	var end = data_off + trace.id_len
	var md5 = trace.md5_array
	append_double_q()
	for i := data_off; i < end; i++ {
		append_result(TRACE_DATA[i])
	}
	append_double_q()
	append_colon()
	append_double_q()
	for i := 0; i < 16; i++ {
		var b = (md5[i]) & 0xff
		append_result(hexes[b>>4])
		append_result(hexes[b&0xf])
	}
	append_double_q()
	append_comma()
}

func getTrace(trace_id int64) *Trace {
	var trace = TRACE_MAP[trace_id]
	if trace == nil {
		trace = &(Trace{})
		trace.init(trace_id)
		TRACE_MAP[trace_id] = trace
	}
	return trace
}

type Trace struct {
	p         int
	data_off  int
	pt_off    int
	id_len    int
	size      int
	data_size int
	md5_array [16]byte
}

func (t *Trace) init(trace_id int64) {
	if t.id_len == 0 {
		t.p = trace_index
		t.pt_off = t.p * MAX_ERR_TRACE
		t.data_off = t.p * MAX_TRACE_LEN * MAX_ERR_TRACE
		t.id_len = get_id_len(trace_id)
		trace_index++
	}
}

func (t *Trace) md5(_MD5 *hash.Hash) {
	var MD5 = *_MD5
	MD5.Reset()
	var pt_off = t.pt_off
	var pt_end = pt_off + t.size
	var array Int64s = TRACE_TIME_ARRAY[pt_off:pt_end]
	sort.Sort(array)
	for i := pt_off; i < pt_end; i++ {
		var raw_time = TRACE_TIME_ARRAY[i]
		var index = get_trace_index(raw_time)
		var raw_pos = TRACE_POS_ARRAY[pt_off+index]
		var pos = get_trace_pos(raw_pos)
		var len = get_trace_len(raw_pos)
		var td_slice = TRACE_DATA[t.data_off+pos : t.data_off+pos+len]
		MD5.Write(td_slice)
	}
	var res = (MD5.Sum(nil))
	copy(t.md5_array[0:16], res)
}

func (t *Trace) write(buf []byte, MD5 *hash.Hash) {
	var pt_off = t.pt_off
	var data_off = t.data_off
	var data_size = t.data_size
	var i = t.size

	var buf_len = len(buf)
	var buf_off = 0

	for ; buf_off < buf_len; {
		var len = int(binary.LittleEndian.Uint16(buf[buf_off : buf_off+2]))
		buf_off += 2
		copy(TRACE_DATA[data_off+data_size:data_off+data_size+len], buf[buf_off:buf_off+len])
		buf_off += len
		var time_off = data_off + data_size + t.id_len + 1
		var time = get_trace_time_from_trace_line(TRACE_DATA[time_off : time_off+16])
		_, exists := TIME_SET.Get(time)
		if exists {
			continue
		} else {
			TIME_SET.Put(time, 1)
		}
		TRACE_TIME_ARRAY[pt_off+i] = to_trace_time(time, i)
		TRACE_POS_ARRAY[pt_off+i] = to_trace_pos(data_size, len)
		data_size += len
		i++
	}
	t.size = i
	t.data_size = data_size
	t.md5(MD5)
	if DEBUG {
		if t.size > MAX_ERR_TRACE {
			log_and_exit("err_trace_overflow")
		}
	}
}

func append_eq() {
	append_result('%')
	append_result('3')
	append_result('D')
}

func append_left_brace() {
	append_result('%')
	append_result('7')
	append_result('B')
}

func append_double_q() {
	append_result('%')
	append_result('2')
	append_result('2')
}

func append_colon() {
	append_result('%')
	append_result('3')
	append_result('A')
}

func append_right_brace() {
	append_result('%')
	append_result('7')
	append_result('D')
}

func append_comma() {
	append_result('%')
	append_result('2')
	append_result('C')
}

func append_result(b byte) {
	RESULT[result_write_index] = b
	result_write_index++
}

func get_trace_time_from_trace_line(data []byte) int64 {
	var end = len(data)
	var time int64 = 0
	for i := 0; i < end; i++ {
		time = time*10 + int64(data[i]-'0')
	}
	return time
}

type Int64s []int64

func (a Int64s) Len() int {
	return len(a)
}

func (a Int64s) Less(i, j int) bool {
	return a[i] < a[j]
}

func (a Int64s) Swap(i, j int) {
	a[i], a[j] = a[j], a[i]
}

func get_id_len(trace_id int64) int {
	var len = ID_SKIP
	trace_id = trace_id >> uint64(len*4)
	for true {
		if trace_id > 0 {
			len++
		} else {
			return len
		}
		trace_id >>= 4
	}
	return -1
}

func to_trace_pos(pos int, len int) uint32 {
	return uint32((pos << 16) | len)
}

func get_trace_pos(raw_pos uint32) int {
	return int(raw_pos >> 16)
}

func get_trace_len(raw_pos uint32) int {
	return int(raw_pos & 0xffff)
}

func to_trace_time(time int64, index int) int64 {
	return (time << 7) | int64(index)
}

func get_trace_time_from_raw_time(raw_time uint64) uint64 {
	return raw_time >> 7
}

func get_trace_index(raw_time int64) int {
	return int(raw_time & (0xff >> 1))
}
